// @axelar-network/axelarjs-sdk/dist/src/services/test/RestService.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const RestService_1 = require("../RestService");
const cross_fetch_1 = __importDefault(require("cross-fetch"));
const mockedFetch = cross_fetch_1.default;
jest.mock("cross-fetch", () => {
    // Mock the default export
    return {
        __esModule: true,
        default: jest.fn(),
    };
});
describe("RestService", () => {
    const host = "http://localhost:3000";
    const api = new RestService_1.RestService(host);
    beforeEach(() => {
        jest.clearAllMocks();
    });
    describe("execRest()", () => {
        describe("when error", () => {
            describe("when text response", () => {
                beforeAll(() => {
                    mockedFetch.mockRejectedValue({
                        text: () => "hello world",
                    });
                });
                it("should return error", () => __awaiter(void 0, void 0, void 0, function* () {
                    yield expect(api.get("/")).rejects.toMatchObject({
                        message: "AxelarJS-SDK uncaught post error",
                        uncaught: true,
                        fullMessage: "hello world",
                    });
                }));
            });
            describe("when json response", () => {
                beforeAll(() => {
                    mockedFetch.mockRejectedValue({
                        json: () => ({
                            message: "Forbidden",
                        }),
                    });
                });
                it("should return error", () => __awaiter(void 0, void 0, void 0, function* () {
                    expect(api.get("/")).rejects.toMatchObject({
                        message: "AxelarJS-SDK uncaught post error",
                        uncaught: true,
                        fullMessage: "Forbidden",
                    });
                }));
            });
        });
        describe("when success", () => {
            let res;
            beforeAll(() => __awaiter(void 0, void 0, void 0, function* () {
                mockedFetch.mockResolvedValue({
                    ok: true,
                    json: () => ({ foo: "bar" }),
                });
                res = yield api.get("/");
            }));
            it("should return json", () => {
                expect(res.foo).toEqual("bar");
            });
        });
    });
});
